hbase-server/src/main/java/org/apache/hadoop/hbase/master/store/LocalStore.java

```diff
@@ -0,0 +1,148 @@
+
+import java.io.IOException;
+import java.util.concurrent.TimeUnit;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.Server;
+import org.apache.hadoop.hbase.TableName;
+import org.apache.hadoop.hbase.client.ColumnFamilyDescriptorBuilder;
+import org.apache.hadoop.hbase.client.Get;
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.TableDescriptor;
+import org.apache.hadoop.hbase.client.TableDescriptorBuilder;
+import org.apache.hadoop.hbase.regionserver.HRegion.FlushResult;
+import org.apache.hadoop.hbase.regionserver.RegionScanner;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+
+import org.apache.hbase.thirdparty.com.google.common.annotations.VisibleForTesting;
+
+/**
+ * Used for storing data at master side. The data will be stored in a {@link LocalRegion}.
+ */
+@InterfaceAudience.Private
+public final class LocalStore {
+
+  // Use the character $ to let the log cleaner know that this is not the normal wal file.
+  public static final String ARCHIVED_WAL_SUFFIX = "$masterlocalwal$";
+
+  public static final String ARCHIVED_HFILE_SUFFIX = "$masterlocalhfile$";
+
+  private static final String MAX_WALS_KEY = "hbase.master.store.region.maxwals";
+
+  private static final int DEFAULT_MAX_WALS = 10;
+
+  public static final String USE_HSYNC_KEY = "hbase.master.store.region.wal.hsync";
+
+  public static final String MASTER_STORE_DIR = "MasterData";
+
+  private static final String FLUSH_SIZE_KEY = "hbase.master.store.region.flush.size";
+
+  private static final long DEFAULT_FLUSH_SIZE = TableDescriptorBuilder.DEFAULT_MEMSTORE_FLUSH_SIZE;
+
+  private static final String FLUSH_PER_CHANGES_KEY = "hbase.master.store.region.flush.per.changes";
+
+  private static final long DEFAULT_FLUSH_PER_CHANGES = 1_000_000;
+
+  private static final String FLUSH_INTERVAL_MS_KEY = "hbase.master.store.region.flush.interval.ms";
+
+  // default to flush every 15 minutes, for safety
+  private static final long DEFAULT_FLUSH_INTERVAL_MS = TimeUnit.MINUTES.toMillis(15);
+
+  private static final String COMPACT_MIN_KEY = "hbase.master.store.region.compact.min";
+
+  private static final int DEFAULT_COMPACT_MIN = 4;
+
+  private static final String ROLL_PERIOD_MS_KEY = "hbase.master.store.region.walroll.period.ms";
+
+  private static final long DEFAULT_ROLL_PERIOD_MS = TimeUnit.MINUTES.toMillis(15);
+
+  private static final String RING_BUFFER_SLOT_COUNT = "hbase.master.store.ringbuffer.slot.count";
+
+  private static final int DEFAULT_RING_BUFFER_SLOT_COUNT = 128;
+
+  public static final TableName TABLE_NAME = TableName.valueOf("master:store");
+
+  public static final byte[] PROC_FAMILY = Bytes.toBytes("proc");
+
+  private static final TableDescriptor TABLE_DESC = TableDescriptorBuilder.newBuilder(TABLE_NAME)
+    .setColumnFamily(ColumnFamilyDescriptorBuilder.of(PROC_FAMILY)).build();
```

Comment:
oh, the second family will come in here if we decide to store root table in here?  Not till then. Ok. I think that makes sense. If table already exists, will have to alter it when we add the root table CF?

Reply:
I think just changing the TABLE_DESC here and then restarting is enough. It just works like the old time, where the meta descriptor is hard coded in code.

Reply:
So the plan would be to store multiple responsibilities (proc store, root store, other projects?) in the same `LocalRegion` instance? Each of these responsibilities would get it's own `LocalStore`/column family? Why do this vs. having a dedicated `LocalRegion` instance for each responsibility?

Reply:
Use their own column family. The problems of introduing more local regions are:

1. Each region would have its own fail recovery logic, which means we need to do lease recovery multiple times when starting master.
2. Maybe we need to duplicate all the configuations here multiple times for supporting different type of local regions, which is a pain for operators.
3. Maybe we need to introduce more types of HFileCleaner and WALCleaner, which is also a pain.